source/h1_connection.c

```diff
@@ -221,70 +222,18 @@ static bool s_connection_new_requests_allowed(const struct aws_http_connection *
-
-    /* Validate the response and cache info that encoder will eventually need.
-     * The encoder_message object will be moved into the stream later while holding the lock */
-    struct aws_h1_encoder_message encoder_message;
-    bool body_headers_ignored = h1_stream->base.request_method == AWS_HTTP_METHOD_HEAD;
-    err = aws_h1_encoder_message_init_from_response(&encoder_message, stream->alloc, response, body_headers_ignored);
-    if (err) {
-        send_err = aws_last_error();
-        goto response_error;
-    }
-
-    bool should_schedule_task = false;
-    { /* BEGIN CRITICAL SECTION */
-        aws_h1_connection_lock_synced_data(connection);
-        if (h1_stream->synced_data.has_outgoing_response) {
-            AWS_LOGF_ERROR(AWS_LS_HTTP_CONNECTION, "id=%p: Response already created on the stream", (void *)stream);
-            send_err = AWS_ERROR_INVALID_STATE;
-        } else {
-            h1_stream->synced_data.has_outgoing_response = true;
-            h1_stream->encoder_message = encoder_message;
-            if (encoder_message.has_connection_close_header) {
-                /* This will be the last stream connection will process, new streams will be rejected */
-                h1_stream->is_final_stream = true;
-                connection->synced_data.new_stream_error_code = AWS_ERROR_HTTP_CONNECTION_CLOSED;
-            }
-
-            if (!connection->synced_data.is_outgoing_stream_task_active) {
-                connection->synced_data.is_outgoing_stream_task_active = true;
-                should_schedule_task = true;
-            }
-        }
-        aws_h1_connection_unlock_synced_data(connection);
-    } /* END CRITICAL SECTION */
-
-    if (send_err) {
-        goto response_error;
-    }
-
-    /* Success! */
-    AWS_LOGF_DEBUG(
-        AWS_LS_HTTP_STREAM, "id=%p: Created response on connection=%p: ", (void *)stream, (void *)connection);
-
-    if (should_schedule_task) {
-        AWS_LOGF_TRACE(AWS_LS_HTTP_CONNECTION, "id=%p: Scheduling outgoing stream task.", (void *)&connection->base);
-        aws_channel_schedule_task_now(connection->base.channel_slot->channel, &connection->outgoing_stream_task);
-    }
-
-    return AWS_OP_SUCCESS;
+    return aws_h1_stream_send_response(h1_stream, response);
+}
 
-response_error:
-    aws_h1_encoder_message_clean_up(&encoder_message);
-    aws_raise_error(send_err);
-    AWS_LOGF_ERROR(
-        AWS_LS_HTTP_STREAM,
-        "id=%p: Sending response on the stream failed, error %d (%s)",
-        (void *)stream,
-        aws_last_error(),
-        aws_error_name(aws_last_error()));
-    return AWS_OP_ERR;
+/* Schedule the cross-thread work task.
+ * `connection->synced_data.is_cross_thread_work_task_scheduled` must be set to true
+ * (while holding the lock) before calling this function. Do NOT call this if
+ * `is_cross_thread_work_task_scheduled` was ALREADY true */
```

Comment:
Yes, it's very brittle, but I don't think the alternatives are better.

Alternative 1: This function grabs the lock and checks whether the task should be scheduled. Downside: Many public API calls involve 2 lock() calls instead of 1.

Alternative 2: Call this function while holding the lock. It flips the bool and schedules the task if necessary. Downside: If the task runs synchronously (ex: scheduling channel task during channel shutdown) then we're still holding the lock, and that might lead to callbacks firing that also try to grab the lock.

I could make the function name ridiculously explicit like `s_schedule_cross_thread_work_task_after_flipping_bool_true()`
seriously

Reply:
Removed function.
Just manually scheduling the task from the same functions that dealt with the locking minutiae